### PR TITLE
Fix chromecast behind reverse proxy

### LINKED ISSUE #3712: Chromecast Broken in Reversy Proxy configuration
**System (please complete the following information):**
 - OS: Windows
 - Virtualization:none
 - Clients: Browser
 - Browser: Chrome 84.0.4147.89 
 - Jellyfin Version: 10.6
 - Playback: Stream to Chromecast
 - Installed Plugins: Pushbullet
 - Reverse Proxy: NGINX
 - Base URL: no
 - Networking: see attached
 - Storage: local

I have/had a working configuration with 10.5.5 using JF to cast to my Chromecast devices, whether those were located internal to my network, or whether it was remote from the JF to another local Chromecast device (i.e. local to my caster/phone).

I run NGINX as RP in what I would consider a mostly standard configuration.  The only non-standard thing about my setup is that my ISP blocks port 80, so I am unable to get any http traffic.  I don't see this as a complication however for clients since most RP proxy users are looking to block all unsecured traffic anyway.  The only reason it has any effect for me, is I am unable to redirect requests that might hit http:// to https:// since no http:// will ever even hit my proxy.  This simply means all my clients must be instructed to point directly to https://

I access my server both internally and externally over:  https://jellyfin.mydomain.com.
All SSL certs are on the NGINX server, and installing them locally on JF doesn't sound like a solution.

I have attached the relevant Network settings from my 10.5.5 server.  These are the only boxes I have filled in and this configuration works properly for all casting in 10.5.5.  As soon as the upgrade to 10.6 I am unable to cast.  I can choose the cast receiver, but when I try to press play on the media absolutely nothing happens (it's as if I can't even click the play button).

While I have reverted to 10.5.5 since you may require working logs as well I don't have complete 10.6 logs of the issue.  However it appears I get messages like when I run in console:

`Error processing request: Unable to find the specified file. URL: http://jellyfin.mydomain.com/socket?`

I believe that this is a problem, since JF is trying to access my site over http even though I am accessing it over https.  I have seen errors like this in multiple places with 10.6, including when trying to get Pushbullet plugin installed and working.

I don't believe that JF should have to require any http communication when working in a RP setting, but it is clear from the network settings that the RP settings have been removed and I see no other configuration that might control this.

Jellyfin 10 5 5 -network-settings

Please let me know what you would like me to gather on 10.5.5 and also on 10.6 after an upgrade to help you diagnose.

### PR DESCRIPTION
**Pull request by jyte**
**Changes**

Fix for /System/Endpoint so it does not return `IsInNetwork: true` when it shouldn't. 

**Issues**

Fixes #3712

### CODE REVIEW

**Comment by BaronGreenback** [comment]
The code works as it does a reverse lookup on host name then checks it against the network.

**Comment by joshuaboniface** [comment]
Please target PRs *only* against `master` branch.

**Review by joshuaboniface — COMMENTED** [review]
I can't yet comment on the completeness of this, since I haven't reviewed the surrounding code, but I would like to see some examples of the structures here to demonstrate (a) how the existing value causes the issue, and (b) how this fixes this. To my eyes right now this looks to just relate to Base URL situations?

**Comment by cvium** [comment]
This is not a proper fix. The "problem" is that our setup doesn't utilize X-Forwarded-For correctly. I have made a PR to fix this #4116

**Comment by jyte** [comment]
> Didn't even know that this function existed!
> 
> The code works as it does a reverse lookup on host name then checks it against the network.
> 
> Let me look further, but doesn't Request.AbsoluteURI always have a value - in which case it will never use remoteIp and always result in a dns lookup.

I do not know if Request.AbsoluteURI always have a value, but it might be possible that that value looks like "http://192.168.0.x/..." in which case there won't be any DNS lookup.



> I can't yet comment on the completeness of this, since I haven't reviewed the surrounding code, but I would like to see some examples of the structures here to demonstrate (a) how the existing value causes the issue, and (b) how this fixes this. To my eyes right now this looks to just relate to Base URL situations?

First of, `request.Endpoint` is null, (no idea how or if it can be populated), so the system fallback to `Request.RemoteIp`

The thing is, when using Nginx as reverse proxy, the Request.RemoteIp is not the actual client IP, but the nginx server IP. So if Nginx is on the same LAN as the JF server, `_network.IsInLocalNetwork` will always return true, and this is wrong because the real client is on a different network (might even come from Internet). 

Reading the code of `_network.IsInLocalNetwork`, it can deal with hostname, ip or even plain url. It does IP parsing, hostname lookup etc. The point being : is the real client (and not the reverse proxy) in the same network as the JF instance ?

Further down the road, in jellyfin-web code for src/plugins/chromecastPlayer/chromecastHelper.js there is this :
```js
return apiClient.getEndpointInfo().then(function (endpoint) {
        if (endpoint.IsInNetwork) {
            return apiClient.getPublicSystemInfo().then(function (info) {
                let localAddress = info.LocalAddress;
                if (!localAddress) {
                    console.debug('No valid local address returned, defaulting to external one');
                    localAddress = serverAddress;
                }
                addToCache(serverAddress, localAddress);
                return localAddress;
            });
        } else {
            addToCache(serverAddress, serverAddress);
            return serverAddress;
        }
    });
```
`serverAddress` is the web url. So according to this, when the JF server is in the same network, we use its local address, but when it's on a different network, we just keep the current URL. 
With the current bug, the client think it is on the same network as the JF instance, and will send to the chromecast the JF server localaddress, which is not reachable because in fact it really is a different (unreachable) network.

**Comment by jyte** [comment]
> This is not a proper fix. The "problem" is that our setup doesn't utilize X-Forwarded-For correctly. I have made a PR to fix this #4116

I thought of the X-Forwarded-For thing and talked about it in the issue #3712 . Debugging made me realise that this solution was less change and would perfectly work too regarding the chromecast.

**Comment by cvium** [comment]
Your fix only works coincidentally because the request IP/hostname is _usually_ not a LAN IP. I won't set up a test to verify this, but I believe if someone set up another proxy in the correct LAN subnet and forwarded to your proxy, the url would be incorrectly marked as being in the local network.

Either way, I don't agree with the fix and the network code we have is pretty awful, but rewriting it is still somewhat far in the future. My PR is the standard way to handle client IP extraction from proxies. Unfortunately it can't be backported, so I am not sure what to do about 10.6.

**Comment by BaronGreenback** [comment]
@cvium - checked your PR. Yes it will sort things, not only in this instance, but throughout the rest of the code.

**Comment by cvium** [comment]
My fix is only needed for unstable/10.7.

I am not sure why 10.6 is broken. If X-Forwarded-For is set correctly `IsInNetwork: false` is returned from my testing.

**Comment by BaronGreenback** [comment]
I found a place in the 10.6 code where the X-Forward headers weren't present - can't remember where.

**Comment by jyte** [comment]
I agree there are still some issues with that PR, like if the JF server can resolve the hostname to a different IP than the client and if that resolution end up on being some computer on the JF LAN by setting some values in the hosts file for instance.

Still, I think this improve the situation to many possible setup (not all), and bring back chrome cast functionnality to current stable branch for people like me. I for instance cannot downgrade to 10.5 and 10.6 is broken. 

I agree that your PR is better overall approach, but my aim was to get something light that could be included in next 10.6 release (I have been missing that funtionnality long enough) without having to wait for the new 10.7 which have many changes regarding network config and will likely have plenty of other problem (I am not saying you are doing bad, just that it is what usually happens when major rewrites are done)

**Comment by cvium** [comment]
Yes, we break stuff all the time. Such is the cost of progress.

For 10.6 I am more interested in why X-Forwarded-For isn't working. I did a simple CURL to /System/Endpoint and it correctly sets `IsInNetwork` based on the header.

**Comment by jyte** [comment]
I have single proxy, initial reporter of the issue also have single proxy if I got it right.

My setup is : 
192.168.0.1 internet router (forward port 80 & 443 to 192.168.0.17)
192.168.0.17 (server hosting internal NAT) port 443 & 80 forwarded to it
----10.13.221.107 (lxd container) nginx reverse proxy (port 80 & 443 forwarded from host to the container)
----10.13.221.198 (lxd container) Jellyfin

192.168.0.11 (computer i cast from using 'jellyfin.mydomain.fr' that resolve to my router public IP)
192.168.0.15 (chromecast)

**Comment by jyte** [comment]
I cannot do testing now, I'll report on curl results latter. But based on what you posted, since you forward for `127.0.0.1` I would have expected `IsLocal` to be `true`

**Comment by jyte** [comment]
I fail to see how `X-Forwarded-For` would have any impact on 10.6.z branch. The only place I can find any mention of that header is in the `WebSocketSharpRequest.cs` which (looking at the code) seems to process the header well. But that would only be for websocket connection, not for a simple GET request.

I tried anyway, and whatever I put in `X-Forwarded-For` has no impact. RemoteIP is the reverse proxy IP when I curl through the RP, or the computer IP I curl from if I go straight to JF server.

**Comment by jyte** [comment]
All right, mystery solved. Problem is not on jellyfin. When given the right ip in X-Forwarded-For, it correctly picks it up.

What should happen : 
- I connect to my domain, 
- the request goes to the router, 
- remote IP become the router public IP, 
- router realise it is its own ip
- send the request straight to nginx
- nginx adds headers X-Forwarded-For with the remote_ip
- nginx forward the query with the new headers to JF
- Jellyfin /System/Endpoint "IsInNetwork" is false because the router public IP in X-Forwarded-For is not anykind of private IP (regardless of whether the private IP is on the same network of the JF server or not).

And this actually works when the nginx server is not inside an lxd container.

The problem comes when the nginx is an LXD container with port forwarding from host to container setup the "normal easy way". (I actually have no idea if there is anyother way to bind host port to container than the one causing me issue, yet). 
Basically, the remote_ip nginx sets in X-Forwarded-For is 127.0.0.1.
It seems to be some kind of known issue ([i found this](https://discuss.linuxcontainers.org/t/how-to-get-real-client-ip-when-using-lxd-to-forward-port-80/2079)), and not specific to LXD but more like to how containers in general handle some network things...

I guess you can discard the PR then. 

Thanks for all the help & insight for helping me troubleshooting this.

**Comment by cvium** [comment]
@jyte Not sure I understand that thread. Have you tried adding this to your nginx config?
```
real_ip_header X-Forwarded-For;
set_real_ip_from 127.0.0.1;
```

**Comment by jyte** [comment]
I just did, here is what it looks like : 
```
location / {
        # Proxy main Jellyfin traffic
        proxy_pass http://jellyfin:8096/;
        proxy_set_header Host $host;
        # proxy_set_header X-Real-IP $remote_addr;
        # proxy_set_header X-Real-IP 8.8.8.8;
        # proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        # proxy_set_header X-Forwarded-For 8.8.8.8;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Forwarded-Protocol $scheme;
        proxy_set_header X-Forwarded-Host $http_host;
        real_ip_header X-Forwarded-For;
        set_real_ip_from 127.0.0.1;


        # Disable buffering when the nginx proxy gets very resource heavy upon streaming
        proxy_buffering off;
    }
```

And here is the related tcpdump output on the JF server 
```
GET /System/Endpoint HTTP/1.0
Host: jellyfin.mydomain.tld
X-Forwarded-Proto: https
X-Forwarded-Protocol: https
X-Forwarded-Host: jellyfin.mydomain.tld
Connection: close
Accept-Encoding: deflate, gzip
authority: jellyfin.mydomain.tld
pragma: no-cache
cache-control: no-cache
accept: application/json
user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.135 Safari/537.36
x-emby-authorization: MediaBrowser Client="Jellyfin Web", Device="Chrome", DeviceId="...i edited that...", Version="10.6.4", Token="629de2e6378a4c77b27bce4d0368eec6"
sec-fetch-site: same-origin
sec-fetch-mode: cors
sec-fetch-dest: empty
referer: https://jellyfin.mydomain.tld/web/index.html
accept-language: fr-FR,fr;q=0.9,en-US;q=0.8,en;q=0.7
```

It's missing the X-Forwarded-For header

**Comment by cvium** [comment]
@jyte You're not using it correctly. It's from the nginx realip module, which you may not have. You don't use it with `proxy_set_header`. Alternatively, if you're feeling adventurous you can install openresty and then use this to make nginx spit out the entire request as-is, which should make it easier to see what you have available.
```
http {
    access_log off;
    include mime.types;

    server {
        listen 8080;
        listen [::]:8080 ipv6only=on;
        location / {
          echo_duplicate 1 $echo_client_request_headers;
          echo "\r";
          echo_read_request_body;
          echo $request_body;
        }
    }
}
```
